network/p2p/dns/resolver_test.go

```diff
@@ -201,6 +202,59 @@ func TestResolver_Expired_Invalidated(t *testing.T) {
 	require.Zero(t, txtSize)
 }
 
+// TestResolver_OverloadQueue initializes a resolver with worker = 1 and queue size = 2 for both IPLookup and TXTLookup
+// Test will then attempt to overload the queue to test that queries outside the queue size would be dropped but resolver does not fail.
+func TestResolver_OverloadQueue(t *testing.T) {
+	ipQueryCount := 0
+	txtQueryCount := 0
+	basicResolver := mocknetwork.BasicResolver{}
+	dnsCache := herocache.NewDNSCache(
+		DefaultCacheSize,
+		unittest.Logger(),
+		metrics.NewNoopCollector(),
+		metrics.NewNoopCollector(),
+	)
+
+	resolver := NewResolver(
+		unittest.Logger(),
+		metrics.NewNoopCollector(),
+		dnsCache,
+		WithBasicResolver(&basicResolver),
+		WithTTL(1*time.Second),
+		WithResolverSize(1, 1, 2, 2))
+
+	cancelCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx, _ := irrecoverable.WithSignaler(cancelCtx)
+	resolver.Start(ctx)
+	unittest.RequireCloseBefore(t, resolver.Ready(), 100*time.Millisecond, "could not start dns resolver on time")
+
+	size := 10 // 10 text and 10 ip domains.
+	times := 1 // each domain is queried 5 times.
+	txtTestCases := testnetwork.TxtLookupFixture(size)
+	ipTestCases := testnetwork.IpLookupFixture(size)
+
+	// each domain is resolved only once through the underlying resolver, and then is cached for subsequent times.
+	mockBasicResolverForDomains(t, &basicResolver, ipTestCases, txtTestCases, happyPath, 2, LogResolveQueryCounts(&ipQueryCount, &txtQueryCount))
```

Comment:
While this approach looks correct with passing the counters by parameters, I still recommend evaluating the correctness of the test through mocking the metrics and then asserting the expected calls have been done. 

Hence, please take the approach suggested in my previous comment: https://github.com/onflow/flow-go/pull/2771#discussion_r918085619

Compared to the current state of implementation, the suggested metrics-based approach adds the following values: 
1. It makes the test simpler, consistent with the existing ones and more readable, maintainable, and flexible. 
2. It does not need adding two more variables as well as dealing with their pointer-based lifecycle. 
3. It does not need to add more complexities to the test helpers, i.e., `mockBasicResolverForDomain`. 
4. It also tests the correctness of our metrics collection procedure for the cache.